framework/audio/internal/mixer.cpp

```diff
@@ -0,0 +1,242 @@
+}
+
+std::shared_ptr<IAudioInsert> Mixer::insert(unsigned int number) const
+{
+    IF_ASSERT_FAILED(m_insertList.find(number) != m_insertList.end()) {
+        return nullptr;
+    }
+    return m_insertList.at(number);
+}
+
+void Mixer::setInsert(unsigned int number, std::shared_ptr<IAudioInsert> insert)
+{
+    IF_ASSERT_FAILED(insert->streamCount() == streamCount()) {
+        LOGE() << "Insert's stream count not equal to the channel";
+        return;
+    }
+    m_insertList[number] = insert;
+}
+
+unsigned int Mixer::addChannel(std::shared_ptr<IAudioSource> source)
+{
+    auto guard = std::lock_guard(m_process);
+    auto lastId = (m_inputList.size() > 0 ? m_inputList.rbegin()->first : -1);
+    auto newId = lastId + 1;
+
+    auto channel = std::make_shared<MixerChannel>();
+    channel->setSource(source);
+    channel->setBufferSize(m_buffer.size());
+    channel->setSampleRate(m_sampleRate);
+
+    m_inputList[newId] = channel;
+    return newId;
+}
+
+void Mixer::removeChannel(unsigned int channelId)
+{
+    auto guard = std::lock_guard(m_process);
+    m_inputList.erase(channelId);
+}
+
+void Mixer::setActive(unsigned int channelId, bool active)
+{
+    auto guard = std::lock_guard(m_process);
+    m_inputList[channelId]->setActive(active);
+}
+
+void Mixer::setLevel(unsigned int channelId, unsigned int streamId, float level)
+{
+    auto guard = std::lock_guard(m_process);
+    m_inputList[channelId]->setLevel(streamId, level);
+}
+
+void Mixer::setBalance(unsigned int channelId, unsigned int streamId, std::complex<float> balance)
+{
+    auto guard = std::lock_guard(m_process);
+    m_inputList[channelId]->setBalance(streamId, balance);
+}
+
+std::shared_ptr<IMixerChannel> Mixer::channel(unsigned int number) const
+{
+    IF_ASSERT_FAILED(m_inputList.find(number) != m_inputList.end()) {
+        return nullptr;
+    }
+    return m_inputList.at(number);
+}
+
+void Mixer::setBufferSize(unsigned int samples)
+{
+    IAudioSource::setBufferSize(samples);
+    for (auto& input : m_inputList) {
+        input.second->setBufferSize(samples);
+    }
+}
+
+void Mixer::forward(unsigned int sampleCount)
+{
+    auto guard = std::lock_guard(m_process);
+    std::fill(m_buffer.begin(), m_buffer.end(), 0.f);
+
+    if (m_clock) {
+        m_clock->forward(sampleCount);
+    }
+
+    for (auto& input : m_inputList) {
+        input.second->forward(sampleCount);
+        mixinChannel(input.second, sampleCount);
+    }
+
+    for (auto& insert : m_insertList) {
+        if (insert.second->active()) {
+            insert.second->process(m_buffer.data(), m_buffer.data(), sampleCount);
+        }
+    }
+    std::transform(m_buffer.begin(), m_buffer.end(), m_buffer.begin(),
+                   [this](float sample) -> float { return sample * m_masterLevel; });
+}
+
+void Mixer::buildRpcReflection()
+{
+    using namespace rpc;
+    m_rpcReflection = {
+        {
+            "setActive",  [this](ArgumentList args) -> Variable {
```

Comment:
I believe, what this class should not know about RPC system (types, signatures and etc)

Reply:
fixed